source/core_mqtt_serializer.c

```diff
@@ -357,95 +588,1308 @@ static MQTTStatus_t checkPublishRemainingLength( size_t remainingLength,
+    {
+        const uint8_t * pVariableHeader = *pIndex;
+        MQTTStatus_t status = MQTTSuccess;
+
+        /*Protocol error to include the same property twice.*/
+        if( *pUsed == true )
+        {
+            status = MQTTProtocolError;
+        }
+        /*Validate the length and decode.*/
+        else if( *pPropertyLength < sizeof( uint32_t ) )
+        {
+            status = MQTTMalformedPacket;
+        }
+        else
+        {
+            *pProperty = UINT32_DECODE( pVariableHeader );
+            pVariableHeader = &pVariableHeader[ sizeof( uint32_t ) ];
+            *pUsed = true;
+            *pPropertyLength -= sizeof( uint32_t );
+        }
+
+        *pIndex = pVariableHeader;
+        return status;
+    }
+
+    static MQTTStatus_t decodeuint16_t( uint16_t * pProperty,
+                                        size_t * pPropertyLength,
+                                        bool * pUsed,
+                                        const uint8_t ** pIndex )
+    {
+        const uint8_t * pVariableHeader = *pIndex;
+        MQTTStatus_t status = MQTTSuccess;
+
+        /*Protocol error to include the same property twice.*/
+
+        if( *pUsed == true )
+        {
+            status = MQTTProtocolError;
+        }
+        /*Validate the length and decode.*/
+
+        else if( *pPropertyLength < sizeof( uint16_t ) )
+        {
+            status = MQTTMalformedPacket;
+        }
+        else
+        {
+            *pProperty = UINT16_DECODE( pVariableHeader );
+            pVariableHeader = &pVariableHeader[ sizeof( uint16_t ) ];
+            *pUsed = true;
+            *pPropertyLength -= sizeof( uint16_t );
+        }
+
+        *pIndex = pVariableHeader;
+        return status;
+    }
+
+    static MQTTStatus_t decodeuint8_t( uint8_t * pProperty,
+                                       size_t * pPropertyLength,
+                                       bool * pUsed,
+                                       const uint8_t ** pIndex )
+    {
+        const uint8_t * pVariableHeader = *pIndex;
+        MQTTStatus_t status = MQTTSuccess;
+
+        /*Protocol error to include the same property twice.*/
+
+        if( *pUsed == true )
+        {
+            status = MQTTProtocolError;
+        }
+        /*Validate the length and decode.*/
+
+        else if( *pPropertyLength < sizeof( uint8_t ) )
+        {
+            status = MQTTMalformedPacket;
+        }
+        else
+        {
+            *pProperty = *pVariableHeader;
+            pVariableHeader = &pVariableHeader[ sizeof( uint8_t ) ];
+            *pUsed = true;
+            *pPropertyLength -= sizeof( uint8_t );
+
+            if( *pProperty > 1U )
+            {
+                status = MQTTProtocolError;
+            }
+        }
+
+        *pIndex = pVariableHeader;
+        return status;
+    }
+
+    static MQTTStatus_t decodeutf_8( const char ** pProperty,
+                                     uint16_t * pLength,
```

Comment:
Why we need the extra argument of pLength in this decode function while not in others. If we really need "pLength", then will the check in Line 1699 be against "pLength" instead of "sizeof( uint16_t )"

Reply:
It is used to store the length of the UTF 8 string. Before decoding, property length is compared with 2 byte for validation and then length is decoded.